hal/src/nRF52840/watchdog_hal.cpp

```diff
@@ -0,0 +1,213 @@
+
+class WatchdogLock {
+public:
+    WatchdogLock() {
+        mutex_.lock();
+    }
+
+    ~WatchdogLock() {
+        mutex_.unlock();
+    }
+
+private:
+    StaticRecursiveMutex mutex_;
+};
+
+/**
+ * @brief The watchdog reset behavior is the same as the pin reset behavior.
+ * The following resources will be reset due to watchdog reset:
+ *  - CPU
+ *  - Peripherals
+ *  - GPIO
+ *  - Debug
+ *  - RAM
+ *  - WDT
+ *  - Retained registers
+ * 
+ * The watchdog will be reset by the following reset sources:
+ *  - Watchdog reset
+ *  - Pin reset
+ *  - Brownout reset
+ *  - Power on reset
+ * 
+ * @note When the device starts running again, after a reset, or waking up from OFF mode,
+ * the watchdog configuration registers will be available for configuration again.
+ */
+class Nrf52Watchdog : public Watchdog {
+public:
+    int init(const hal_watchdog_config_t* config) {
+        CHECK_FALSE(initialized_, SYSTEM_ERROR_INVALID_STATE);
+        CHECK_TRUE(config && (config->size > 0), SYSTEM_ERROR_INVALID_ARGUMENT);
+        CHECK_TRUE(config->timeout_ms >= minTimeout(), SYSTEM_ERROR_INVALID_ARGUMENT);
+        CHECK_TRUE(config->timeout_ms <= maxTimeout(), SYSTEM_ERROR_INVALID_ARGUMENT);
+
+        nrfx_wdt_config_t nrfConfig = {
+            .behaviour          = NRF_WDT_BEHAVIOUR_PAUSE_SLEEP_HALT, // WDT will be paused when CPU is in SLEEP or HALT mode.
+            .reload_value       = config->timeout_ms,
+            .interrupt_priority = NRF52_WATCHDOG_PRIORITY,
+        };
+        nrfx_err_t ret = nrfx_wdt_init(&nrfConfig, nrf52WatchdogEventHandler);
+        SPARK_ASSERT(ret == NRF_SUCCESS);
+        ret = nrfx_wdt_channel_alloc(&channelId_);
+        SPARK_ASSERT(ret == NRF_SUCCESS);
+
+        memcpy(&config_, config, std::min(config_.size, config->size));
+        initialized_ = true;
+        return SYSTEM_ERROR_NONE;
+    }
+
+    int start() {
+        CHECK_TRUE(initialized_, SYSTEM_ERROR_INVALID_STATE);
+        CHECK_FALSE(started(), SYSTEM_ERROR_NONE);
+        nrfx_wdt_enable();
+        CHECK_TRUE(started(), SYSTEM_ERROR_INTERNAL);
+        return SYSTEM_ERROR_NONE;
+    }
+
+    bool started() override {
+        info_.running = nrf_wdt_started();
+        return info_.running;
+    }
+
+    int refresh() {
+        CHECK_TRUE(initialized_, SYSTEM_ERROR_INVALID_STATE);
+        CHECK_TRUE(started(), SYSTEM_ERROR_INVALID_STATE);
+        nrfx_wdt_channel_feed(channelId_);
+        return SYSTEM_ERROR_NONE;
+    }
+
+    int setOnExpiredCallback(hal_watchdog_on_expired_callback_t callback, void* context) override {
+        // NOTE: The max amount of time we can spend in WDT interrupt is two cycles of 32768[Hz] clock - after that, reset occurs
+        callback_ = callback;
+        context_ = context;
+        return SYSTEM_ERROR_NONE;
+    }
+
+    static Nrf52Watchdog* instance() {
+        static Nrf52Watchdog watchdog(WATCHDOG_CAPS_INT, minTimeout(), maxTimeout());
```

Comment:
`CAPS_RESET` as well, as the caller needs to understand that while we get a notification nothing we can do about not resetting.